README
This is the README for App-Addex.  It uses your address book to produce config
for various unixy mail tools.


Changes
Revision history for App-Addex

{{$NEXT}}

0.027     2021-06-19 13:28:38-04:00 America/New_York
        - update author contact info
        - add perl-support section

0.026     2014-03-31 21:41:23-04:00 America/New_York
        - use phrase/addr-spec output for muttrc instead of comment form
          (recent mutts seem to use comments if you use parents in muttrc
          aliases, which is no good)

0.025     2013-10-24 22:34:16 America/New_York
          fix a bunch of duplicated documentation

0.024     2013-07-23 11:21:27 America/Los_Angeles
          fix a boneheaded precedence error in the Mutt plugin; thanks, Reini
          Urban!

          repackage with Dist::Zilla

0.023     2012-02-27
          sort entries before processing

0.022     2010-06-13
          updated for new Config::MVP

0.021     2009-07-27
0.020     2009-07-27
          updated for new Config::MVP

0.019     2009-01-16
          update packaging a bit

0.018     2008-06-22
          make ->entries a delegated method
          make Output::Mutt not give (name) for groups; mutt pukes

0.017     2008-05-29
          most of Config has been refactored to Config::INI::MVP::Reader
          use that!

0.016     2008-04-13
          fix missing prereq

0.015     2008-04-12
          Output::ToFile now opens output filehandles to write utf8
          add unidecode option for Output::Mutt
          output all label/alias combinations for Output::Mutt

0.014     2008-02-17
          AddressBook objects can now refer back to the App::Addex object
          experimental "plugin" classes

          EmailAddress objects may be "send" or "receive" or both now

0.013     2007-12-17
          declare that we require 5.008; easier than fixing tests, but patches
            welcome, I guess

0.012     2007-11-12
          stop brokenly reaching into guts of Config::INI::Reader object
            (thanks, JCAP!)
          remove spurious 'use' statement in addex program

0.011     2007-11-11
          use Mixin::ExtraFields::Param for ->field method (dogfood!)
          remove spurious Mac::Glue prereq

0.010     2007-11-05
          add base Output class to make future expansion easier
          add finalize output method
          suppress some warnings during testing

0.009     2007-10-24
          upgrade Module::Install

0.008     2007-05-11
          wrote tests for just about everything
          fix mutt labels to start at with 1 after the first of each label

0.007     2007-05-11
          App::Addex::Entry::EmailAddress
          mutt output includes address labels in aliases if available

0.006     2007-05-10
          introduce App::Addex::Config, making config files much easier

0.005     2007-05-09
          fixed a really stupid, broken test

0.004     2007-05-09
          ensure that new plugins are getting hashref args

0.003     2007-05-05
          change procmail output to use . for folder paths

0.001     2007-03-21
          first release


lib/App/Addex.pm
use strict;
use warnings;

# because tests use open to open a string-ref, and I am not interested in ever
# supporting this module on ancient perls -- rjbs, 2007-12-17
use 5.008;

package App::Addex;
# ABSTRACT: generate mail tool configuration from an address book

use Carp ();

=head1 DESCRIPTION

B<Achtung!>  The API to this code may very well change.  It is almost certain
to be broken into smaller pieces, to support alternate sources of entries, and
it might just get plugins.

This module iterates through all the entries in an address book and produces
configuration file based on the entries in the address book, using configured
output plugins.

It is meant to be run with the F<addex> command, which is bundled as part of
this software distribution.

=method new

  my $addex = App::Addex->new(\%arg);

This method returns a new Addex.

Valid parameters are:

  classes    - a hashref of plugin/class pairs, described below

Valid keys for the F<classes> parameter are:

  addressbook - the App::Addex::AddressBook subclass to use (required)
  output      - an array of output producers (required)

For each class given, an entry in C<%arg> may be given, which will be used to
initialize the plugin before use.

=cut

# sub starting_section_name { 'classes' }
sub mvp_multivalue_args  { qw(output plugin) }

sub new {
  my ($class, $arg) = @_;

  my $self = bless {} => $class;

  # XXX: keep track of seen/unseen classes; carp if some go unused?
  # -- rjbs, 2007-04-06

  for my $core (qw(addressbook)) {
    my $class = $arg->{classes}{$core}
      or Carp::confess "no $core class provided";

    $self->{$core} = $self->_initialize_plugin($class, $arg->{$class});
  }

  my @output_classes = @{ $arg->{classes}{output} || [] }
    or Carp::confess "no output classes provided";

  my @output_plugins;
  for my $class (@output_classes) {
    push @output_plugins, $self->_initialize_plugin($class, $arg->{$class});
  }
  $self->{output} = \@output_plugins;

  my @plugin_classes = @{ $arg->{classes}{plugin} || [] };
  for my $class (@plugin_classes) {
    eval "require $class" or die;
    $class->import(%{ $arg->{$class} || {} });
  }

  return $self;
}

sub from_sequence {
  my ($class, $seq) = @_;

  my %arg;
  for my $section ($seq->sections) {
    $arg{ $section->name } = $section->payload;
  }

  $class->new(\%arg);
}

sub _initialize_plugin {
  my ($self, $class, $arg) = @_;
  $arg ||= {};
  $arg->{addex} = $self;

  # in most cases, this won't be needed, since the App::Addex::Config will have
  # loaded plugins as a side effect, but let's be cautious -- rjbs, 2007-05-10
  eval "require $class" or die;

  return $class->new($arg);
}

=method addressbook

  my $abook = $addex->addressbook;

This method returns the App::Addex::AddressBook object.

=cut

sub addressbook { $_[0]->{addressbook} }

=method output_plugins

This method returns all of the output plugin objects.

=cut

sub output_plugins {
  my ($self) = @_;
  return @{ $self->{output} };
}

=method entries

This method returns all the entries to be processed.  By default it is
delegated to the address book object.  This method may change a good bit in the
future, as we really want an iterator, not just a list.

=cut

sub entries {
  my ($self) = @_;
  return sort { $a->name cmp $b->name } $self->addressbook->entries;
}

=method run

  App::Addex->new({ ... })->run;

This method performs all the work expected of an Addex: it iterates through the
entries, invoking the output plugins for each one.

=cut

sub run {
  my ($self) = @_;

  for my $entry ($self->entries) {
    for my $plugin ($self->output_plugins) {
      $plugin->process_entry($self, $entry);
    }
  }

  for my $plugin ($self->output_plugins) {
    $plugin->finalize;
  }
}

1;
